File: backend/app/api/v1/photos.py
"""
Body-zone catalog + patient photo gallery API.

Photos are uploaded as multipart, persisted via services.photo_storage
(local fs in dev; swap to S3 in prod), and served back through an
authenticated GET endpoint so consent + tenant isolation are enforced
on every read.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile, status
from fastapi.responses import StreamingResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.patient import Patient
from app.models.photo import (
    BodyZone,
    PatientPhoto,
    PhotoAngle,
    PhotoConsentScope,
    PhotoStage,
    ZoneCategory,
)
from app.models.user import User
from app.schemas.photo import (
    BodyZoneCreate,
    BodyZoneListResponse,
    BodyZoneResponse,
    PhotoListResponse,
    PhotoResponse,
)
from app.services import photo_storage
from app.services.zone_seed import seed_zones_for_clinic


router = APIRouter()


def _get_clinic_id(user: User) -> uuid.UUID:
    membership = next((m for m in user.memberships if m.is_active), None)
    if not membership:
        raise HTTPException(status_code=403, detail="No active clinic membership")
    return membership.clinic_id


# ---------- Zones -------------------------------------------------------

@router.get("/zones", response_model=BodyZoneListResponse)
async def list_zones(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    clinic_id = _get_clinic_id(user)
    res = await db.execute(
        select(BodyZone)
        .where(BodyZone.clinic_id == clinic_id, BodyZone.is_active.is_(True))
        .order_by(BodyZone.sort_order.asc(), BodyZone.name_fr.asc())
    )
    rows = list(res.scalars().all())
    return BodyZoneListResponse(
        zones=[BodyZoneResponse.model_validate(r) for r in rows],
        total=len(rows),
    )


@router.post("/zones/seed")
async def seed_zones(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    clinic_id = _get_clinic_id(user)
    inserted = await seed_zones_for_clinic(db, clinic_id)
    return {"inserted": inserted}


@router.post("/zones", response_model=BodyZoneResponse, status_code=status.HTTP_201_CREATED)
async def create_zone(
    body: BodyZoneCreate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    clinic_id = _get_clinic_id(user)
    try:
        cat = ZoneCategory(body.category)
    except ValueError:
        raise HTTPException(status_code=400, detail=f"Unknown category: {body.category}")
    z = BodyZone(
        clinic_id=clinic_id,
        slug=body.slug,
        name_fr=body.name_fr,
        name_ar=body.name_ar,
        name_en=body.name_en,
        category=cat,
        sort_order=body.sort_order,
    )
    db.add(z)
    await db.commit()
    await db.refresh(z)
    return BodyZoneResponse.model_validate(z)


# ---------- Photos ------------------------------------------------------

@router.get("", response_model=PhotoListResponse)
async def list_photos(
    patient_id: uuid.UUID,
    zone_slug: str | None = None,
    stage: str | None = None,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    clinic_id = _get_clinic_id(user)
    stmt = (
        select(PatientPhoto)
        .where(
            PatientPhoto.clinic_id == clinic_id,
            PatientPhoto.patient_id == patient_id,
            PatientPhoto.deleted_at.is_(None),
        )
        .order_by(PatientPhoto.captured_at.desc())
    )
    if zone_slug:
        stmt = stmt.where(PatientPhoto.zone_slug == zone_slug)
    if stage:
        try:
            stmt = stmt.where(PatientPhoto.stage == PhotoStage(stage))
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Unknown stage: {stage}")
    res = await db.execute(stmt)
    rows = list(res.scalars().all())
    return PhotoListResponse(
        photos=[PhotoResponse.model_validate(r) for r in rows],
        total=len(rows),
    )


@router.post("/upload", response_model=PhotoResponse, status_code=status.HTTP_201_CREATED)
async def upload_photo(
    patient_id: uuid.UUID = Form(...),
    zone_slug: str = Form(...),
    stage: str = Form("before"),
    angle: str | None = Form(None),
    consent_scope: str = Form("medical"),
    plan_id: uuid.UUID | None = Form(None),
    appointment_id: uuid.UUID | None = Form(None),
    note: str | None = Form(None),
    file: UploadFile = File(...),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    clinic_id = _get_clinic_id(user)

    # Validate patient
    pres = await db.execute(
        select(Patient).where(Patient.id == patient_id, Patient.clinic_id == clinic_id)
    )
    if not pres.scalar_one_or_none():
        raise HTTPException(status_code=404, detail="Patient not found")

    # Validate zone
    zres = await db.execute(
        select(BodyZone).where(BodyZone.clinic_id == clinic_id, BodyZone.slug == zone_slug)
    )
    if not zres.scalar_one_or_none():
        raise HTTPException(status_code=400, detail=f"Unknown zone: {zone_slug}")

    # Validate enums
    try:
        stage_enum = PhotoStage(stage)
    except ValueError:
        raise HTTPException(status_code=400, detail=f"Unknown stage: {stage}")
    angle_enum: PhotoAngle | None = None
    if angle:
        try:
            angle_enum = PhotoAngle(angle)
        except ValueError:
            raise HTTPException(status_code=400, detail=f"Unknown angle: {angle}")
    try:
        consent_enum = PhotoConsentScope(consent_scope)
    except ValueError:
        raise HTTPException(status_code=400, detail=f"Unknown consent_scope: {consent_scope}")

    # Read + persist
    data = await file.read()
    if not data:
        raise HTTPException(status_code=400, detail="Empty upload")
    if len(data) > 25 * 1024 * 1024:
        raise HTTPException(status_code=413, detail="File too large (max 25 MB)")

    ext = (file.filename or "").rsplit(".", 1)[-1] if file.filename and "." in file.filename else "jpg"
    storage_key, size_bytes = photo_storage.save(
        clinic_id=clinic_id,
        patient_id=patient_id,
        ext=ext,
        data=data,
    )

    photo = PatientPhoto(
        clinic_id=clinic_id,
        patient_id=patient_id,
        plan_id=plan_id,
        appointment_id=appointment_id,
        captured_by=user.id,
        storage="local",
        storage_key=storage_key,
        content_type=file.content_type or "image/jpeg",
        size_bytes=size_bytes,
        zone_slug=zone_slug,
        stage=stage_enum,
        angle=angle_enum,
        consent_scope=consent_enum,
        captured_at=datetime.now(timezone.utc),
        note=note,
    )
    db.add(photo)
    await db.commit()
    await db.refresh(photo)
    return PhotoResponse.model_validate(photo)


@router.get("/{photo_id}/file")
async def serve_photo_file(
    photo_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    clinic_id = _get_clinic_id(user)
    res = await db.execute(
        select(PatientPhoto).where(
            PatientPhoto.id == photo_id,
            PatientPhoto.clinic_id == clinic_id,
            PatientPhoto.deleted_at.is_(None),
        )
    )
    photo = res.scalar_one_or_none()
    if not photo:
        raise HTTPException(status_code=404, detail="Photo not found")

    try:
        data = photo_storage.open_bytes(photo.storage_key)
    except FileNotFoundError:
        raise HTTPException(status_code=404, detail="Photo file missing on disk")

    return StreamingResponse(
        iter([data]),
        media_type=photo.content_type or "image/jpeg",
        headers={"Cache-Control": "private, max-age=300"},
    )


@router.delete("/{photo_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_photo(
    photo_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    clinic_id = _get_clinic_id(user)
    res = await db.execute(
        select(PatientPhoto).where(
            PatientPhoto.id == photo_id,
            PatientPhoto.clinic_id == clinic_id,
            PatientPhoto.deleted_at.is_(None),
        )
    )
    photo = res.scalar_one_or_none()
    if not photo:
        raise HTTPException(status_code=404, detail="Photo not found")
    photo.deleted_at = datetime.now(timezone.utc)
    await db.commit()
    return None

File: frontend/lib/api/photos.ts
import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import { apiClient } from "@/lib/api-client";
import { useAuthStore } from "@/stores/auth-store";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

// ── Types ──────────────────────────────────────────────────────

export type PhotoStage = "before" | "during" | "after" | "follow_up" | "control";
export type PhotoAngle =
  | "front" | "left_45" | "right_45" | "left_profile" | "right_profile" | "back" | "detail" | "other";
export type ConsentScope = "medical" | "before_after" | "marketing";

export interface BodyZone {
  id: string;
  slug: string;
  name_fr: string;
  name_ar: string | null;
  name_en: string | null;
  category: "face" | "body" | "hair" | "extremities";
  sort_order: number;
  is_active: boolean;
}

export interface PatientPhoto {
  id: string;
  patient_id: string;
  plan_id: string | null;
  appointment_id: string | null;
  captured_by: string | null;
  storage: string;
  storage_key: string;
  content_type: string;
  width: number | null;
  height: number | null;
  size_bytes: number | null;
  zone_slug: string;
  stage: PhotoStage;
  angle: PhotoAngle | null;
  consent_scope: ConsentScope;
  captured_at: string;
  note: string | null;
  created_at: string;
}

// ── Helpers ────────────────────────────────────────────────────

export function photoFileUrl(photoId: string): string {
  return `${API_BASE_URL}/api/v1/photos/${photoId}/file`;
}

// ── Hooks ──────────────────────────────────────────────────────

export function useBodyZones() {
  const token = useAuthStore((s) => s.accessToken);
  return useQuery({
    queryKey: ["photos", "zones"],
    queryFn: () =>
      apiClient<{ zones: BodyZone[]; total: number }>("/photos/zones", {
        token: token ?? undefined,
      }),
  });
}

export function useSeedZones() {
  const token = useAuthStore((s) => s.accessToken);
  const qc = useQueryClient();
  return useMutation({
    mutationFn: () =>
      apiClient<{ inserted: number }>("/photos/zones/seed", {
        method: "POST",
        token: token ?? undefined,
      }),
    onSuccess: () => qc.invalidateQueries({ queryKey: ["photos", "zones"] }),
  });
}

export function usePatientPhotos(patientId: string | undefined) {
  const token = useAuthStore((s) => s.accessToken);
  return useQuery({
    queryKey: ["photos", "patient", patientId],
    queryFn: () =>
      apiClient<{ photos: PatientPhoto[]; total: number }>(
        `/photos?patient_id=${patientId}`,
        { token: token ?? undefined }
      ),
    enabled: !!patientId,
  });
}

export function useUploadPhoto() {
  const token = useAuthStore((s) => s.accessToken);
  const qc = useQueryClient();

  return useMutation({
    mutationFn: async (input: {
      file: File;
      patient_id: string;
      zone_slug: string;
      stage: PhotoStage;
      angle?: PhotoAngle;
      consent_scope?: ConsentScope;
      plan_id?: string;
      appointment_id?: string;
      note?: string;
    }) => {
      const fd = new FormData();
      fd.append("file", input.file);
      fd.append("patient_id", input.patient_id);
      fd.append("zone_slug", input.zone_slug);
      fd.append("stage", input.stage);
      if (input.angle) fd.append("angle", input.angle);
      if (input.consent_scope) fd.append("consent_scope", input.consent_scope);
      if (input.plan_id) fd.append("plan_id", input.plan_id);
      if (input.appointment_id) fd.append("appointment_id", input.appointment_id);
      if (input.note) fd.append("note", input.note);

      const res = await fetch(`${API_BASE_URL}/api/v1/photos/upload`, {
        method: "POST",
        headers: token ? { Authorization: `Bearer ${token}` } : {},
        body: fd,
      });
      if (!res.ok) {
        const text = await res.text();
        throw new Error(text || `HTTP ${res.status}`);
      }
      return (await res.json()) as PatientPhoto;
    },
    onSuccess: (p) => {
      qc.invalidateQueries({ queryKey: ["photos", "patient", p.patient_id] });
    },
  });
}

export function useDeletePhoto(patientId: string) {
  const token = useAuthStore((s) => s.accessToken);
  const qc = useQueryClient();
  return useMutation({
    mutationFn: async (photoId: string) => {
      const res = await fetch(`${API_BASE_URL}/api/v1/photos/${photoId}`, {
        method: "DELETE",
        headers: token ? { Authorization: `Bearer ${token}` } : {},
      });
      if (!res.ok) throw new Error(`HTTP ${res.status}`);
    },
    onSuccess: () =>
      qc.invalidateQueries({ queryKey: ["photos", "patient", patientId] }),
  });
}
